### PR TITLE
[Rule] Ban transitive dependencies

### PR DESCRIPTION
**Pull request by johnwiseheart**
Ban transitive dependencies.

I added this as a new rule (instead of as an addition to `bannedDependencies`) as I chose to parse `yarn.lock` instead of running through the dependency tree for each individual package. Thus this rule would be a "root rule" (like `nestedWorkspaces`) and not applied for each package in a monorepo. Happy to go that route if it's preferred, but this way was easier.

### CODE REVIEW

File: packages/rules/src/__tests__/bannedTransitiveDependencies.spec.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
+/*!
```
Comment (styu) [inline packages/rules/src/__tests__/bannedTransitiveDependencies.spec.ts]: delete this header?

File: packages/rules/src/bannedTransitiveDependencies.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+export const bannedTransitiveDependencies: RuleModule<typeof Options> = {
+  check: (context: Context, opts: Options) => {
+    const { bannedTransitiveDependencies: bannedDeps } = opts;
+
+    const packagePath = context.getWorkspaceContext().packageDir;
+    const yarnLockFilePath = path.join(packagePath, "yarn.lock");
+
+    const yarnLock: LockFileObject = lockfileParse(readFileSync(yarnLockFilePath).toString()).object;
+    const allDeps = Object.keys(yarnLock);
```
Comment (styu) [inline packages/rules/src/bannedTransitiveDependencies.ts]: we only care about direct deps though right? e.g. if i pull in a banned dep as a dev dep for types only, do we care?

Comment (johnwiseheart) [reply packages/rules/src/bannedTransitiveDependencies.ts]: I suppose so, but parsing yarn.lock its not possible to know what is and isn't dev dependencies right? We'd have to do a dependency tree parse for that (I naively tried a `npm ls X` solution but it was way too slow). Might need to pull some of Walker's DependencyTreeService out.

Comment (styu) [reply packages/rules/src/bannedTransitiveDependencies.ts]: hm looking at yarn.lock i don't think transitive dev deps show up in yarn.lock, but dev deps of package.jsons in the repo do. i think in that case you would need to traverse the in-repo dependency tree and filter things out, similar to the some of the internal logic we have

Comment (johnwiseheart) [reply packages/rules/src/bannedTransitiveDependencies.ts]: Resolved - not using yarn.lock any more for this

File: tslint.json [hunk 3]
```diff
@@ -6,8 +6,8 @@
     "file-header": [
       true,
 
-      "Copyright 2019 Palantir Technologies, Inc\\.",
-      "Copyright 2019 Palantir Technologies, Inc.\n\nLicensed under the MIT license. See LICENSE file in the project root for details.\n"
+      "Copyright \\d{4} Palantir Technologies, Inc\\.",
+      "Copyright 2020 Palantir Technologies, Inc.\n\nLicensed under the MIT license. See LICENSE file in the project root for details.\n"
```
Comment (styu) [inline tslint.json]: why isn't this one `\\d{4}` ?

Comment (johnwiseheart) [reply tslint.json]: Its the default for autofixing

**Pull request merged**